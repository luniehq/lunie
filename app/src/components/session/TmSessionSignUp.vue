<template>
  <SessionFrame icon="person_add">
    <TmFormStruct :submit="onSubmit">
      <h2 class="session-title">Create a new address</h2>
      <div
        v-if="
          !isExtension && !session.insecureMode && !session.mobile && !isTestnet
        "
        class="session-main"
      >
        <InsecureModeWarning />
      </div>
      <div v-else>
        <div class="session-main bottom-indent">
          <Steps
            v-if="!session.mobile"
            :steps="[`Create`, `Password`, `Backup`]"
            active-step="Create"
          />
          <DangerZoneWarning v-if="!session.mobile" />
          <TmFormGroup
            :error="$v.fieldName.$error"
            field-id="sign-up-name"
            field-label="Account Name"
          >
            <TmField
              id="sign-up-name"
              v-model.trim="fieldName"
              v-focus
              type="text"
              placeholder="Must be at least 3 characters"
              vue-focus="vue-focus"
            />
            <TmFormMsg
              v-if="$v.fieldName.$error && !$v.fieldName.required"
              name="Name"
              type="required"
            />
            <TmFormMsg
              v-if="$v.fieldName.$error && !$v.fieldName.minLength"
              name="Name"
              type="minLength"
              min="3"
            />
            <TmFormMsg
              v-if="$v.fieldName.$error && !$v.fieldName.nameExists"
              name="Name"
              type="custom"
              msg="already exists"
            />
          </TmFormGroup>
        </div>
        <div class="session-footer">
          <TmBtn value="Next" type="submit" />
        </div>
      </div>
    </TmFormStruct>
  </SessionFrame>
</template>

<script>
import { mapState, mapGetters } from "vuex"
import { required, minLength } from "vuelidate/lib/validators"
import TmBtn from "common/TmBtn"
import TmFormGroup from "common/TmFormGroup"
import TmFormStruct from "common/TmFormStruct"
import TmField from "common/TmField"
import TmFormMsg from "common/TmFormMsg"
import SessionFrame from "common/SessionFrame"
import DangerZoneWarning from "common/DangerZoneWarning"
import InsecureModeWarning from "common/InsecureModeWarning"
import Steps from "../../ActionModal/components/Steps"
import { getWalletIndex } from "@lunie/cosmos-keys"

const nameExists = (value) => {
  const walletIndex = getWalletIndex()
  if (walletIndex.some((e) => e.name === value)) {
    return false
  } else {
    return true
  }
}

export default {
  name: `session-sign-up`,
  components: {
    TmBtn,
    TmField,
    SessionFrame,
    TmFormGroup,
    TmFormMsg,
    TmFormStruct,
    DangerZoneWarning,
    InsecureModeWarning,
    Steps,
  },
  computed: {
    ...mapState([`session`, `signup`]),
    ...mapGetters([`network`, `networks`, `isExtension`]),
    fieldName: {
      get() {
        return this.$store.state.signup.signUpName
      },
      set(value) {
        this.$store.commit(`updateField`, { field: `signUpName`, value })
      },
    },
    isTestnet() {
      const selectedNetwork = this.networks.find(
        ({ id }) => id === this.network
      )
      return selectedNetwork ? selectedNetwork.testnet : false
    },
  },
  methods: {
    async onSubmit() {
      this.$v.$touch()
      if (this.$v.$error) return
      this.$router.push(`/create/password`)
    },
  },
  validations: () => ({
    fieldName: { required, minLength: minLength(3), nameExists },
  }),
}
</script>
