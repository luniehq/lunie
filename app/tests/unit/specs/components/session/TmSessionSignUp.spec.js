import { createLocalVue, shallowMount } from "@vue/test-utils"
import Vuelidate from "vuelidate"
import TmSessionSignUp from "session/TmSessionSignUp"
jest.mock("@lunie/cosmos-keys", () => ({
  getWalletIndex: function () {
    return [{ name: `Happy Lunie User`, address: `xyz123` }]
  },
}))

describe(`TmSessionSignUp`, () => {
  const localVue = createLocalVue()
  localVue.directive(`focus`, () => {})
  localVue.use(Vuelidate)

  let wrapper, $store

  beforeEach(() => {
    $store = {
      state: {
        session: { insecureMode: true },
        signup: {
          signUpName: "",
        },
      },
      commit: jest.fn(),
      dispatch: jest.fn(),
      mutations: {
        updateField: jest.fn(),
      },
      getters: {
        network: `cosmos-hub-mainnet`,
        isExtension: false,
      },
    }

    wrapper = shallowMount(TmSessionSignUp, {
      localVue,
      mocks: {
        $store,
        $router: {
          push: jest.fn(),
        },
      },
      stubs: [`router-link`],
    })
  })

  it("renders", () => {
    expect(wrapper.element).toMatchSnapshot()
  })

  it(`should commit updateField on field change`, async () => {
    wrapper.setData({ fieldName: `HappyLunieUser` })
    expect(wrapper.vm.$store.commit).toHaveBeenCalledWith(`updateField`, {
      field: `signUpName`,
      value: `HappyLunieUser`,
    })
  })

  it(`validation should fail if name is not filled in`, async () => {
    await wrapper.vm.onSubmit()
    expect(wrapper.vm.$v.fieldName.$error).toBe(true)
  })

  it(`validation should fail if name lenght < 3 characters`, async () => {
    wrapper.vm.$store.state.signup.signUpName = `as`
    await wrapper.vm.onSubmit()
    expect(wrapper.vm.$v.fieldName.$error).toBe(true)
  })

  it(`validation should not fail if name lenght >= 3 characters`, async () => {
    wrapper.vm.$store.state.signup.signUpName = `Happy Lunie User 2`
    await wrapper.vm.onSubmit()
    expect(wrapper.vm.$v.fieldName.$error).toBe(false)
  })

  it(`validation should fail if name exists already in stored accounts`, async () => {
    wrapper.vm.$store.state.signup.signUpName = `Happy Lunie User`
    await wrapper.vm.onSubmit()
    expect(wrapper.vm.$v.fieldName.$error).toBe(true)
  })

  it(`validation should fail if name exists already in stored accounts`, async () => {
    wrapper.vm.$store.state.signup.signUpName = `Happy Lunie User`
    await wrapper.vm.onSubmit()
    expect(wrapper.vm.$v.fieldName.$error).toBe(true)
  })

  it(`should go to /create/password when submit the form`, async () => {
    wrapper.vm.$store.state.signup.signUpName = `HappyLunieUser`
    await wrapper.vm.onSubmit()
    expect(wrapper.vm.$router.push).toHaveBeenCalledWith(`/create/password`)
  })

  it(`isTesnet should return false if current network is mainnet`, () => {
    const self = {
      networks: [
        {
          id: `cosmos-hub-mainnet`,
          testnet: false,
        },
      ],
    }
    const isTesnet = TmSessionSignUp.computed.isTestnet.call(self)
    expect(isTesnet).toBe(false)
  })
})
