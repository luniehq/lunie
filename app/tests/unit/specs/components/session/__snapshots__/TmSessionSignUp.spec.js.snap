// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`TmSessionSignUp renders 1`] = `
<sessionframe-stub
  icon="person_add"
>
  <tmformstruct-stub
    submit="function () { [native code] }"
    width=""
  >
    <h2
      class="session-title"
    >
      Create a new address
    </h2>
     
    <div>
      <div
        class="session-main bottom-indent"
      >
        <steps-stub
          activestep="Create"
          steps="Create,Password,Backup"
        />
         
        <dangerzonewarning-stub />
         
        <tmformgroup-stub
          fieldid="sign-up-name"
          fieldlabel="Account Name"
        >
          <tmfield-stub
            id="sign-up-name"
            placeholder="Must be at least 3 characters"
            type="text"
            value=""
            vue-focus="vue-focus"
          />
           
          <!---->
           
          <!---->
           
          <!---->
        </tmformgroup-stub>
      </div>
       
      <div
        class="session-footer"
      >
        <tmbtn-stub
          customclass=""
          type="submit"
          value="Next"
        />
      </div>
    </div>
  </tmformstruct-stub>
</sessionframe-stub>
`;
